Extension/src/LanguageServer/client.ts

```diff
@@ -574,6 +598,128 @@ export class DefaultClient implements Client {
+                                    // referencesParams does not match the object used when creating the request, abort it.
+                                    if (params !== referencesParams) {
+                                        reject();
+                                    }
+                                    referencesPromiseReject = reject;
+                                    referencesRequestPending = true;
+                                    this.client.languageClient.sendNotification(FindAllReferencesNotification, params);
+
+                                    // Register a single-fire handler for the reply.
+                                    this.client.references.setResultsCallback(result => {
+                                        if (referencesRequestPending) {
+                                            referencesRequestPending = false;
+                                            let locations: vscode.Location[] = [];
+                                            result.referenceInfos.forEach(referenceInfo => {
+                                                if (referenceInfo.type === refs.ReferenceType.Confirmed) {
+                                                    let uri: vscode.Uri = vscode.Uri.file(referenceInfo.file);
+                                                    let range: vscode.Range = new vscode.Range(referenceInfo.position.line, referenceInfo.position.character, referenceInfo.position.line, referenceInfo.position.character + result.text.length);
+                                                    locations.push(new vscode.Location(uri, range));
+                                                }
+                                            });
+                                            resolve(locations);
+                                        }
+                                    });
+                                });
+                                token.onCancellationRequested(e => {
+                                    if (params === referencesParams) {
+                                        if (referencesRequestPending) {
+                                            referencesRequestPending = false;
+                                            this.client.cancelReferences();
+                                            this.client.sendRequestReferences();
+                                            reject();
+                                        } else {
+                                            referencesParams = null;
+                                        }
+                                    }
+                                });
+                            });
+                        }
+                    }
+
+                    this.disposables.push(vscode.languages.registerReferenceProvider(documentSelector, new FindAllReferencesProvider(this)));
+
+                    class RenameProvider implements vscode.RenameProvider {
```

Comment:
Put RenameProvider in different file?

Reply:
All of these providers need to be in their own files.  I'd like to do that in a subsequent PR, after Insiders.